### ISSUE TITLE
unable to commit node_modules folder due to MAX_PATH issue

### ISSUE THREAD

**Issue by Starystars67** [desktop/desktop#3900]
### Description

System Crash When trying to commit over 6000 files (small files)

### Version & Crash Screen
![image](https://user-images.githubusercontent.com/15389482/35628285-55e405a6-0693-11e8-814e-159889e347de.png)

**GitHub Desktop version:** [version here]

**OS version:** [version here]

### Steps to Reproduce

1. [First step] Unsure im trying to commit my files after i have done some work (6000+) files are waiting to commit.
1. [Second step]
1. [and so on]

**Expected behavior:** [What you expected to happen] the files commit

**Actual behavior:** [What actually happened] it crashes

**Reproduces how often:** [What percentage of the time does it reproduce?] everytime i try to commit this application and its some 6000 files
[2018-01-31.desktop.production.log](https://github.com/desktop/desktop/files/1681999/2018-01-31.desktop.production.log)

**Comment by iAmWillShepherd** [comment]
Are you able to accomplish this commit in the CLI?

Since you're using Node.js, I recommend that you add [this](https://github.com/github/gitignore/blob/master/Node.gitignore) `.gitignore` to your project as a workaround. 

Then execute the following commands
1. `git add .gitignore`
1. `git commit -m "add gitignore"`
1. `git rm --cached .`
1. `git add .`
1. `git commit -m "track non-ignored files"`

Going to close this out for now. Feel free to re-open if those steps do not work out for you.

**Comment by shiftkey** [comment]
Just to add to the advice from @iAmWillShepherd, the log file shows Git returning an error:

```
2018-01-31T01:48:01.879Z - error: [ui] `git update-index --add --remove --replace -z --stdin` exited with an unexpected code: 128.
...
The file will have its original line endings in your working directory.
error: lstat("node_modules/grunt/node_modules/dateformat/node_modules/meow/node_modules/normalize-package-data/node_modules/validate-npm-package-license/node_modules/spdx-correct/node_modules/spdx-license-ids/spdx-license-ids.json"): Filename too long
fatal: Unable to process path node_modules/grunt/node_modules/dateformat/node_modules/meow/node_modules/normalize-package-data/node_modules/validate-npm-package-license/node_modules/spdx-correct/node_modules/spdx-license-ids/spdx-license-ids.json
```

This often occurs when you encounter nested `node_modules` folders that exceed the Windows `MAX_PATH` limits. If @iAmWillShepherd's recommendation to ignore these files isn't suitable, you could try moving the repository closer to the root of the drive.

I'd also love to reproduce this issue myself as it feels related to #2041 which I've been chasing for months:

```
2018-01-31T01:48:01.877Z - error: [main] Error: write EOF
    at exports._errnoException (util.js:1050:11)
    at WriteWrap.afterWrite (net.js:813:14)
```

I think these three bits of information will help with reproducing the issue on our end:

 - the contents of your `package.json` contents when making the failed commit
 - the version of Node and NPM you were running at the time
  
```
$ node -v
$ npm -v
```

 - the full path to your repository on disk, so I can create it at the same location and trigger the `MAX_PATH` issue as that might be related

**Comment by Starystars67** [comment]
Hello, Okay so @iAmWillShepherd you suggestion did not work though creating the gitignore did commit and push. @shiftkey Your suggestion of moving it to the root of a drive also has not worked with the same result as my screen shot shown above. i have attached my node and npm versions along with the contents of package.json
Node: v8.9.4
NPM: v5.6.0
Package.json:
{
  "name": "YourThought.Games.NodeJS",
  "version": "1.0.0",
  "description": "Initial Test of moving over to NodeJS",
  "main": "server.js",
  "scripts": {
    "start-dev": "nodemon server.js",
    "start": "node server.js"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/Starystars67/YourThought.Games.NodeJS.git"
  },
  "author": "Mitch Durso",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/Starystars67/YourThought.Games.NodeJS/issues"
  },
  "homepage": "https://github.com/Starystars67/YourThought.Games.NodeJS",
  "dependencies": {
    "body-parser": "^1.18.2",
    "bootstrap": "^4.0.0",
    "cookie-parser": "^1.4.3",
    "debug": "~2.6.9",
    "express": "^4.15.5",
    "express-php": "0.0.1",
    "express-session": "^1.15.6",
    "fontawesome": "^4.7.2",
    "morgan": "~1.9.0",
    "multer": "^1.3.0",
    "mysql": "^2.15.0",
    "mysqli": "^2.0.1",
    "nodemon": "^1.14.11",
    "php-express": "0.0.3",
    "pug": "2.0.0-beta11",
    "serve-favicon": "~2.4.5"
  },
  "devDependencies": {
    "grunt": "^1.0.1",
    "grunt-contrib-jshint": "^1.1.0",
    "nodemon": "^1.14.11"
  }
}

**Comment by iAmWillShepherd** [comment]
@Starystars67 

To expand on @shiftkey's point, Windows is complaining about the file path being too long (see [Maximum Path Length Limitation](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx). Moving it to your root path may not solve it because there is still some potential for the path length to exceed 257 characters.

```
2018-01-31T01:48:01.879Z - error: [ui] `git update-index --add --remove --replace -z --stdin` exited with an unexpected code: 128.
...
The file will have its original line endings in your working directory.
error: lstat("node_modules/grunt/node_modules/dateformat/node_modules/meow/node_modules/normalize-package-data/node_modules/validate-npm-package-license/node_modules/spdx-correct/node_modules/spdx-license-ids/spdx-license-ids.json"): Filename too long
```

The path length of the file causing this error is 217 characters long, so you don't have much wiggle room.

In regards to the steps, I provided not working

1. Did it fail at step 3? If so, what was the error message you received?
1. Are you able to accomplish this commit in the CLI?

**Comment by shiftkey** [comment]
@Starystars67 thanks for the information. There's one more bit that I need to reproduce this:

>  - the full path to your repository on disk, so I can create it at the same location and trigger the `MAX_PATH` issue as that might be related

**Comment by Starystars67** [comment]
![image](https://user-images.githubusercontent.com/15389482/35706987-e12635f0-079f-11e8-81c4-439793bf40de.png)

as you can see i am at the root of my C Drive so it must be one of the node_modules but i have provided a list of them

**Comment by shiftkey** [comment]
@Starystars67 excellent, that will help with ensuring I'm testing this as closely as possible. Thanks!

**Comment by Starystars67** [comment]
No worries, Any Further questions just ask :p

**Comment by shiftkey** [comment]
We made a fix to our embedded Git in https://github.com/desktop/dugite/pull/164 which seems to address the `write EOF` error along with some others, and this is available in `1.1.0` if you update your installs.

If you are still able to trigger this error, please open a new issue with details and we can try and get to the bottom of it.
